x_transformers/multi_IO_transformers.py

```diff
@@ -0,0 +1,495 @@
+                                     range(len(self.post_attn_layers))]
+                if any(dim != self.post_attn_layers[i].dim for i in range(len(self.post_attn_layers))):
+                    print('Note: Since the model dimension is not equal to the output_attn_layers dimension, '
+                          'a linear layer is added to project the model dimension to the output_attn_layers dimension. '
+                          'If this is not desired, please change the model dimensions.')
+                if logits_dim is not None:
+                    if tie_embedding:
+                        assert all(self.post_attn_layers[i].dim == self.post_attn_layers[i].dim for i in range(
+                            len(self.post_attn_layers))), 'if tie_embedding is True, the dimensions of the input and output attn layers must be equal'
+                    self.to_logits = [nn.Linear(dim, d, bias=False) for d in logits_dim] if not tie_embedding else \
+                        lambda t: [t @ self.token_emb[i].emb.weight.t() for i in range(len(logits_dim))]
+                else:
+                    self.to_logits = [nn.Identity() for _ in range(len(self.post_attn_layers))]
+            else:
+                if logits_dim is not None:
+                    if tie_embedding:
+                        self.logits = [lambda t: t @ self.token_emb[i].emb.weight.t() if self.multi_input else lambda t: t @ self.model.token_emb.emb.weight.t() for i in range(len(logits_dim)) ]
+                    else:
+                        self.to_logits = [nn.Linear(dim, d, bias=False) for d in logits_dim]
+                else:
+                    self.to_logits = nn.Identity()
+
+    def init_(self):
+        if self.l2norm_embed:
+            for i in range(len(self.token_emb)):
+                nn.init.normal_(self.token_emb[i].emb.weight, std=1e-5)
+                if not isinstance(self.pos_emb, always):
+                    nn.init.normal_(self.pos_emb[i].emb.weight, std=1e-5)
+        for i in range(len(self.token_emb)):
+            nn.init.kaiming_normal_(self.token_emb[i].emb.weight)
+
+    def forward(
+            self,
+            x,
+            return_embeddings=False,
+            return_logits_and_embeddings=False,
+            return_intermediates=False,
+            mask=None,
+            return_mems=False,
+            return_attn=False,
+            mems=None,
+            mem_masks=None,
+            pos=None,
+            prepend_embeds=None,
+            prepend_mask=None,
+            embed_ids: list[Dict[str, Tensor]] or Dict[str, Tensor] = dict(),
+            sum_embeds=None,
+            return_attn_z_loss=False,
+            attn_z_loss_weight=1e-4,
+            seq_start_pos=None,
+            cache=None,
+            **kwargs
+    ):
+        global intermediates_model, cache_pre_attn_layers, cache_model, cache_post_attn_layers, intermediates_pre_attn_layers, intermediates_pre_attn_layer, mem_packed_shape, mem_every
```

Comment:
are you using a `global` here for the caches?

Reply:
Yes, I don't think it's completely necessary though, I just did it because my IDE wanted me to make them global. It probably could be deleted without causing much issues

Reply:
i see, want to delete it from the PR?